### PR TITLE
feat: introduces Features DSL

### PR DESCRIPTION
**Pull request by bartoszmajsak**
- easy builder to define what particular feature consists of
- enables the use of existing YAML manifests
- allows definition of additional resources programmatically with a simple func registration
- cleanup funcs are now part of each feature and are called one after the other
- each feature has its own resource tracker object

### CODE REVIEW

File: pkg/kfapp/ossm/feature/cert.go
Comment (cam-garrison) [inline pkg/kfapp/ossm/feature/cert.go:1]: Nice, looks like the refactor made things a lot more concise here.

File: pkg/kfapp/ossm/feature/feature.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+		}
+
+		log.Info("applying manifest", "path", m.targetPath())
+	}
+
+	if err := f.applyManifests(); err != nil {
+		return err
+	}
+
+	// TODO postconditions
```
Comment (cam-garrison) [inline pkg/kfapp/ossm/feature/feature.go:83]: Just a question for my own clarification - what would be an example of a postcondition here?

File: pkg/kfapp/ossm/ossm_installer.go [hunk 2]
```diff
@@ -78,187 +75,131 @@ func (o *OssmInstaller) Init(_ kftypesv3.ResourceEnum) error {
...
+		For(o.PluginSpec).
+		UsingConfig(o.config).
+		WithResources(feature.ConfigMaps).
+		Load(); err != nil {
+		return err
+	} else {
+		o.features = append(o.features, cfMaps)
+	}
+
+	if serviceMesh, err := feature.CreateFeature("enable-service-mesh").
```
Comment (cam-garrison) [inline pkg/kfapp/ossm/ossm_installer.go]: nitpick but I think `"meshify-kfdef-ns"` or `"meshify-odh-ns` might be more accurate for this feature name

**Review by cam-garrison — APPROVED** [review]
Only one small nitpick really. Nice :smile:

File: pkg/kfapp/ossm/feature/feature.go (hunk 1 above)
Comment (bartoszmajsak) [reply pkg/kfapp/ossm/feature/feature.go:83]: Wait for the ingress pod to start for example. Probably naming is not the best.

File: pkg/kfapp/ossm/templates/control-plane/base/control-plane-ingress.patch.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    ingress:
+      volumes:
+        - volume:
+            secret:
+              secretName: test-oauth2-tokens
+              optional: true
+          volumeMount:
+            name: test-oauth2-tokens
+            mountPath: "/etc/istio/test-oauth2-tokens"
+            readOnly: true
```
Comment (bartoszmajsak) [inline pkg/kfapp/ossm/templates/control-plane/base/control-plane-ingress.patch.yaml]: Humm.... I think tests are overwriting the actual template now :D

File: pkg/kfapp/ossm/ossm_installer.go (hunk 2 above)
Comment (bartoszmajsak) [reply pkg/kfapp/ossm/ossm_installer.go]: Agree, I did improve those names in 0867d1d

File: pkg/kfapp/ossm/templates/control-plane/base/control-plane-ingress.patch.yaml (hunk 3 above)
Comment (bartoszmajsak) [reply pkg/kfapp/ossm/templates/control-plane/base/control-plane-ingress.patch.yaml]: Fixed in 88224c7

File: pkg/kfapp/ossm/feature/feature.go (hunk 1 above)
Comment (cam-garrison) [reply pkg/kfapp/ossm/feature/feature.go:83]: I think the naming is fine - it clearly refers to some task/action between creation and deletion. I just couldn't think of a good example of one.

File: pkg/kfapp/ossm/templates/control-plane/base/control-plane-ingress.patch.yaml (hunk 3 above)
Comment (cam-garrison) [reply pkg/kfapp/ossm/templates/control-plane/base/control-plane-ingress.patch.yaml]: Ah okay, I was a little confused about why this was added, I should have asked.

**Pull request merged**